Alternative correlation method in indexer.correlate()

EirikOpheim opened the pull request:
---
name: Normalized correlation
about: This pull request adds a second correlation method for indexer correlate(). Both methods have given similar orientation results for a test dataset.

---

**Release Notes**
New feature. Indexer.correlate now takes an additional argument "method". I have labeled the existing correlation metric "FastCorrelation", and the alternative metric "NormalizedCorrelation". "FastCorrelation" is the default value of the method argument. 

**What does this PR do? Please describe and/or link to an open issue.**
This pull request simply adds an alternative correlation metric. If no method argument is given, everything runs as it used to. 

**Describe alternatives you've considered**
I have considered running the IF/Elif statement inside the for loops in order to write fewer lines of code. I believe this would be messy eventually, and also harder to optimize for running speed, as more evaluations would be required. 

**Are there any known issues? Do you need help?**
I have tried to write error messages if a method that is not defined is given as input argument. Please check that this syntax is according to the pyxem standard. Also, I believe that some lines can be removed to make the code cleaner. I am planning to do this in a follow-up PR if this PR is approved. (Remove a few outdated comments, etc.)

pc494 commented:
@EirikOpheim this is a nice start. My first feedback would be you can cut down on quiet a lot of duplication here. Both methods use 
```

 for phase_index, library_entry in enumerate(library.values()):
            orientations = library_entry['orientations']
            pixel_coords = library_entry['pixel_coords']
            intensities = library_entry['intensities']
            # TODO: This is only applicable some of the time, probably use an if + special_local in the for
            pattern_norms = library_entry['pattern_norms']
```
and 
```

                    if corr_local > np.min(corr_saved):
                        or_saved[np.argmin(corr_saved)] = or_local
                        corr_saved[np.argmin(corr_saved)] = corr_local

                    combined_array = np.hstack((or_saved, corr_saved))
                    combined_array = combined_array[np.flip(combined_array[:, 3].argsort())]  # see stackoverflow/2828059 for details
                    top_matches[phase_index, :, 0] = phase_index
                    top_matches[phase_index, :, 2] = combined_array[:, 3]  # correlation
                    for i in np.arange(n_largest):
                        top_matches[phase_index, i, 1] = combined_array[i, :3]  # orientation
```

It's good practice to write things once and only once (and it makes reading pull requests a lot easier) - so what I suggest is you refactor so that all the actually "doing stuff" happens inside these two blocks. In pseudo

```
#set up code, reads library etc
if method == 'fast':
    fast code
elif method =='other'
    other code
#add entry to result etc etc etc
```

Does that make sense?

pc494 commented:
(also this is a pretty good first PR for major content, congrats)

EirikOpheim commented:
Thanks, Phillip! 

I will try and rewrite a bit, so that common chunks are written only once! :)

EirikOpheim commented:
Hello again, 

Are the newest changes now part of the PR? 

I have removed as many duplications as I could. Also, I have removed the check on the method argument in the indexation_generator, because this function only passes this argument on to the indexation_utils function. I thought a clearer indication of where the error is can be given from the function where the argument is actually used, if someone tries an invalid method argument. 

I could not find a pyxem standard for writing error messages, so I went with a simple print + sys.exit() if an invalid argument is used. Is this ok, or should we go with something else? I see that an option would be to define an error class/ set of error classes, but I believe this would have to be a common set of classes for all of pyxem.

pc494 commented:
That was fast! 

I think the "architecture" key that you're missing is 
`select_method_from_method_dict(method, method_dict, **kwargs):`
(in `signals.__init__.py`)

This would involve another refactor (splitting `correlate_library` code out into two functions) but we think it's probably the best way forward as it means you can iterate quickly through new method.

In terms of errors, we tend to raise `ValueError("explanation")`.

EirikOpheim commented:
Hello again, Phillip! 

I appreciate that you have already planned how new methods can be effectively integrated!

I am, however, not quite sure if I understood what you wrote regarding the refactor. Could you give an example / write some pseudocode to underline how the structure should be? 

Should we refactor the current correlate_library into a get_method_dict and a correlate_library. And then the correlate_library function will call the method_dict using select_method_from_method_dict function? 

In that case would it be something like: 
def get_method_dict():
method_dict = {
"Fast" : -Implementation_of_fast(a,b,c) - # a, b and c are parameters 
             - return correlation score

"Normalized": -Implementation_of_normalized(a,b,c,d,e) #a, b, c, d and e are parameters
                       - return correlation score

"Other Method 1": - implementation (..) ... # Method someone would like to add in the future  
                              - return correlation score
}
return method_dict

And then in correlate_library we would have a similar code to the current one, but "If method ==  ... /Elif ==... " will be replaced with correlation_score = select_method_from_method_dict(method, method_dict, **kwargs)?

I have a few follow-up questions if this is how the structure should be, regarding for instance the computation of example parameters d and e, that remain unused if "Fast" is chosen.

pc494 commented:
First, a tip, the <> on the github message bar (the 5th one from the left) gives you code snippets.

In terms of structure, I was thinking that 

`correlate_library()` would work much as it does, except the code inside the if `method==''` would look like

```
if method == 'fast_correlation':
     corr = correlation_score_fast(x,y,**kwargs)
elif method=='normalised_correlation':
    corr = correlation_score_normalized(x,z,**kwargs)
```

EirikOpheim commented:
> First, a tip, the <> on the github message bar (the 5th one from the left) gives you code snippets.

Thanks :) 

So you suggest that I use the select_method_from_method_dict() function only for validating method input, define a function for every method, and pick the method through a series of if/elif statements? 

I believe what I do not completely understand is the role of the select_method_from_method_dict() in this function.

pc494 commented:
Sorry, I didn't make myself very clear. a `method_dict` is a map from a string to a function. So your top line would take say:
`'FastCorrelation'` the string to `fast_correlation` the function. This means you can then call
`method_dict[input_str](**kwargs)` and drop all the if/else

That said, if you write it as I described in my previous comment, I'm happy to do the extra refactor.

EirikOpheim commented:
Ah, I see! Thanks! 

I will try and implement with the method_dict. If things go wrong I might ask for some help with refactoring :)

EirikOpheim commented:
I have now set up the structure with dicts, and confirmed that it is working for both methods, and give a satisfactory (I believe) error message. 

I have not used the function select_method_from_method_dict. It produces a help string on every iteration if I do not give it any kwargs, and it returns a function, not the evaluation of the function. Could it be that we should change the return statement of select_method_from_method_dict to 

`return method_dict[method](**kwargs)`

If we implement this change I believe I can use it in this PR.

I tested writing 

`select_method_from_method_dict(method,method_dict)(arguments) 
`
This works, but prints help(method) on every iteration. 

I believe it would function well if I had written 
`select_method_from_method_dict(method,method_dict,arguments)(arguments) `

But then the arguments in the function call would be only to suppress the print(help(method)) within select_method_from_method_dict.

pc494 commented:
This is good.

I think an improvement would be do move the selection of the method function "up" to `correlate` so that it had the callsign:
```
correlate(self,
                  n_largest=5,
                  mask=None,
                  method = 'FastCorrelation',
                  help=False
                  *args,
                  **kwargs)
```
then change `select_method_from_dict` to have a third argument, `help` (defaults to True). I would then pass the string to `correlate_library` & use if/else statements - but now you don't need to check that the string is sensible.

I'll write a code review to help you through this.

pc494 commented on pyxem/generators/indexation_generator.py:
```diff
@@ -75,6 +77,9 @@ def correlate(self,
             The n orientations with the highest correlation values are returned.
         mask : Array
             Array with the same size as signal (in navigation) or None
+        method : String
```
`str` is preferred to `String`

pc494 commented on pyxem/generators/indexation_generator.py:
```diff
@@ -75,6 +77,9 @@ def correlate(self,
             Array with the same size as signal (in navigation) or None
+        method : String
+            Name of method used to compute correlation between templates and diffraction patterns. Can be
+            'FastCorrelation' or 'NormalizedCorrelation'.
```
pyxem is snake_case, so these should be `'fast_correlation'` and `'normalized_correlation'`

pc494 commented on pyxem/generators/indexation_generator.py:
```diff
@@ -95,29 +100,29 @@ def correlate(self,
             # Index at all real space pixels
             mask = 1
 
```
so I would put the `select_method_from_dict` here, but leave this bit till last.

EirikOpheim replied:
One question regarding this. If I move select_method_from_methods_dict here, I will have to define functions fast_correlation and normalized_correlation within the scope of this file, or import them. Do you suggest I import them from indexation_utils? 

I will then pass the chosen method on to correlate_library as a function? 


` from pyxem.utils.indexation_utils import correlate_library, fast_correlation, normalized_correlation` 

`methods_dict = { "fast_correlation" : fast_correlation,
                     "normalized_correlation" : normalized_correlation
                     }`

`my_method = get_methods_from_methods_dict(method,methods_dict,help)`

Include argument function=my_method in call to correlate_library:
`matches = signal.map(correlate_library,
                             library=library,
                             n_largest=n_largest,
                             method = method, # will still have to provide this, as it will impact which variables to compute.
                             function = my_method,
                             mask=mask,
                             inplace=False,
                             **kwargs)`

pc494 replied:
You should import them from `indexation_utils`, yes. But then pass the string to `.correlate_library`. Doing it this way means a user can get docstrings for the matching metric at the `.correlate` stage **and** we can optimize the internal code (see my review comment at line 130 of `indexation_utils`)

pc494 commented on pyxem/generators/indexation_generator.py:
```diff
@@ -95,29 +100,29 @@ def correlate(self,
+                             n_largest=n_largest,
+                             mask=mask,
+                             inplace=False,
+                             method = method,
```
can we have `method` above `inplace` please. Another boring convention that makes the code easier to read. The arguments that are for `map` come after those for functions

pc494 commented on pyxem/utils/indexation_utils.py:
```diff
@@ -21,12 +21,14 @@
 from itertools import combinations
 import math
 from operator import itemgetter, attrgetter
+import sys
```
This can be deleted?

pc494 commented on pyxem/utils/indexation_utils.py:
```diff
@@ -21,12 +21,14 @@
 from pyxem.utils.expt_utils import _cart2polar
 from pyxem.utils.vector_utils import get_rotation_matrix_between_vectors
 from pyxem.utils.vector_utils import get_angle_cartesian
+from pyxem.signals import select_method_from_method_dict
```
This can also be deleted now.

pc494 commented on pyxem/utils/indexation_utils.py:
```diff
@@ -38,17 +40,39 @@
                                "phase_index rotation_matrix match_rate error_hkls total_error scale center_x center_y".split())
 
+#Functions used in correlate_library.
+def Fast_Correlation(image_intensities,int_local,pn_local, **kwargs):
```
snake_case + docstrings

pc494 commented on pyxem/utils/indexation_utils.py:
```diff
@@ -38,17 +40,39 @@
+    return np.sum(np.multiply(image_intensities, int_local)) / \
+                pn_local  # Correlation is the partially normalized dot product
+
+def Normalized_Correlation(N,image_norm,average_image_intensity,image_intensities,int_local, **kwargs):
```
snake_case + docstrings again.

pc494 commented on pyxem/utils/indexation_utils.py:
```diff
@@ -62,6 +86,9 @@ def correlate_library(image, library, n_largest, mask):
         The number of well correlated simulations to be retained.
     mask : bool
         A mask for navigation axes. 1 indicates positions to be indexed.
+    method : String
```
str, camel_case; added should refer to version not date (ie `0.11.0`)

pc494 commented on pyxem/utils/indexation_utils.py:
```diff
@@ -91,7 +118,18 @@ def correlate_library(image, library, n_largest, mask):
        template matching,” vol. 50, no. 1, pp. 87–99, 2005.
     """
+
+    methods_dict = { 'FastCorrelation' : Fast_Correlation,
```
cut this (we're moving it up a level)

pc494 commented on pyxem/utils/indexation_utils.py:
```diff
@@ -91,7 +118,18 @@ def correlate_library(image, library, n_largest, mask):
+        raise ValueError("Method {} is not defined".format(method))
+
     top_matches = np.empty((len(library), n_largest, 3), dtype='object')
+    N = image.shape[0]*image.shape[1]
```
this should be inside an if clause; if `method=='normalized_correlation'`

pc494 commented on pyxem/utils/indexation_utils.py:
```diff
@@ -104,12 +142,13 @@ def correlate_library(image, library, n_largest, mask):
                 image_intensities = image[px_local[:, 1], px_local[:, 0]]
-                corr_local = np.sum(np.multiply(image_intensities, int_local)) / \
-                    pn_local  # Correlation is the partially normalized dot product
+                corr_local = methods_dict[method](image_intensities = image_intensities, int_local = int_local,
```
if else block is probably best here

pc494 commented on pyxem/utils/indexation_utils.py:
```diff
@@ -38,17 +40,39 @@
+                pn_local  # Correlation is the partially normalized dot product
+
+def Normalized_Correlation(N,image_norm,average_image_intensity,image_intensities,int_local, **kwargs):
+    N_star = len(image_intensities)
```
variables start with lower cases. Also anything that doesn't depend on the library should be done outside of this function, because we run every library entry against an image, so anything that doesn't change can be calculated once rather the N times (N size of library)

EirikOpheim commented:
Phillip,

I merged things into this branch to be up to date with master. I believe this has introduced lacking coverage, as I do not know of anything that should reduce coverage in my code.

pc494 commented:
Nice work! The only* thing I would like changed is rather than passing `chosen_function` through use an if/elif block inside `correlate_library` - otherwise the arguments get very confusing.

Also, now is the time to learn how to write a test! If you have trouble let me know.

* excluding minor changes that I will do myself

pc494 commented:
> Phillip,
> 
> I merged things into this branch to be up to date with master. I believe this has introduced lacking coverage, as I do not know of anything that should reduce coverage in my code.

You introduced new functionality and didn't write a test for it? You can see that that is where the drop has come from on coveralls:
https://coveralls.io/builds/28829445/source?filename=pyxem/utils/indexation_utils.py

pc494 commented:
@EirikOpheim, can you write down what you think the correct formulae for a "Normalized correlation" is (ideally with a source)?

EirikOpheim commented:
> @EirikOpheim, can you write down what you think the correct formulae for a "Normalized correlation" is (ideally with a source)?

Sure! I have referenced an article below the reference for the fast_correlation reference.

pc494 commented:
I'm not sure that' the correlation you want. That method is used for finding a "sub image" within an "image" - we use it (or something similar) to find a diffraction spot in a diffraction pattern. See for example the coins in:
https://stackoverflow.com/questions/48732991/search-for-all-templates-using-scikit-image/52785616#52785616

I believe the intention was to do "full" version of the `(a.b / |a| |b| )` - which is all over the literature, for example:
https://onlinelibrary.wiley.com/doi/epdf/10.1107/S2052520619006723 
(it's from Rauch 2005 obviously, but that's harder to find on google)

hakonanes commented:
I just want to recommend Aimo Winkelmann's (@wiai) discussion of NCC for use with EBSD here: https://xcdskd.readthedocs.io/en/latest/cross_correlation/cross_correlation_coefficient.html

EirikOpheim commented:
I will checkout the article, Håkon!

Phillip, I agree that the procedure in the article I referred to is using template matching in a different way, but I believe that the metric used is still applicable to template matching with a full-image comparison. I will send another article, where it is used in image comparison.  I believe we are talking about Zero-mean normalized cross correlation (ZNCC) and Normalized cross correlation (NCC). Where the article you referred is NCC, and the method I have implemented is ZNCC. 

In the following article, both methods are discussed in terms of use in image recognition:

https://www.ncbi.nlm.nih.gov/pmc/articles/PMC3596837/

"On the other hand, NCC is not invariant under the constant change in brightness. To provide this invariance, Zero-mean Cross-Correlation is used."

I am not sure how important this is for template matching, but I believe ZNCC is a valid metric for comparing templates and diffraction patterns.

EirikOpheim commented:
I believe the Pearson coefficient in the paper Håkon sent is the ZNCC

pc494 commented:
Okay, I think the best route forward here is to make sure that the docstrings for what you've built are clear and accurate, and he can discuss the ins/outs of ZNCC vs NCC at our next meeting. When you're happy with what you've got @EirikOpheim hit the request review and I'll finish this up. 

(also thank you @hakonanes, that did help a lot)

EirikOpheim commented:
I will have another look at the docstrings! I am sorry that the references were a bit sloppy! 

I was wondering if there is a test for the correlate_library function already? In order to test this function for full coverage, I might have to define a library and it would be nice to follow the procedure that was used to test the function before changes were made.

pc494 commented:
> I will have another look at the docstrings! I am sorry that the references were a bit sloppy!
> 
> I was wondering if there is a test for the correlate_library function already? In order to test this function for full coverage, I might have to define a library and it would be nice to follow the procedure that was used to test the function before changes were made.

Everything has a test! That's what a 100% test coverage is (can you tell we're proud of it). In this case it's in https://github.com/pyxem/pyxem/blob/master/pyxem/tests/test_workflows/test_orientation_mapping_phys.py

I'm happy to fix that for you though, because it's a bit of a nuisance.

EirikOpheim commented:
> Everything has a test! That's what a 100% test coverage is (can you tell we're proud of it)

Haha, not bad! :) 

I will try and write the docstrings a bit more clear and make one last commit before I hit the request review button. Should I maybe reference the article from Håkon in place of the current one?

pc494 commented on pyxem/generators/indexation_generator.py:
```diff
@@ -91,33 +102,40 @@ def correlate(self,
-                                 inplace=False,
-                                 **kwargs)
+        chosen_function = select_method_from_method_dict(method,method_dict,print_help)
+        #Not used, but the function tests if selected method is a valid argument, and can print help for selected method.
```
Comments should be above the line they refer to, also I would cut all the words before "test"

pc494 commented on pyxem/signals/__init__.py:
```diff
@@ -49,7 +49,7 @@ def push_metadata_through(dummy, *args, **kwargs):
 
 
-def select_method_from_method_dict(method, method_dict, **kwargs):
+def select_method_from_method_dict(method, method_dict,print_help = True, **kwargs):
```
should have docstrings for `print_help`

pc494 commented on pyxem/signals/__init__.py:
```diff
@@ -75,7 +75,7 @@ def select_method_from_method_dict(method, method_dict, **kwargs):
                                   "See documentation for available "
                                   "implementations.".format(method))
-    elif not kwargs:
+    elif print_help:
```
you weren't to know, but this should be `elif print_help and not kwargs` because of how other bits of the code base behave.

pc494 commented on pyxem/utils/indexation_utils.py:
```diff
@@ -38,18 +38,96 @@
+    return np.sum(np.multiply(image_intensities, int_local)) / \
+                pn_local  # Correlation is the partially normalized dot product
+
+def normalized_correlation(nb_pixels,image_std,average_image_intensity,image_intensities,int_local, **kwargs):
```
I am fairly convinced that this should be called "zero_mean_normalized_correlation" & updates elsewhere to reflect this.

pc494 commented on pyxem/utils/indexation_utils.py:
```diff
@@ -60,9 +138,13 @@ def correlate_library(image, library, n_largest, mask):
         experimental data.
     n_largest : int
         The number of well correlated simulations to be retained.
+    method : String
```
`str` again

pc494 commented on pyxem/utils/indexation_utils.py:
```diff
@@ -90,9 +172,25 @@ def correlate_library(image, library, n_largest, mask):
     E. F. Rauch and L. Dupuy, “Rapid Diffraction Patterns identification through
        template matching,” vol. 50, no. 1, pp. 87–99, 2005.
+
+    K. Briechle and U.D. Hanebeck, "Template Matching using Fast Normalized Cross Correlation"
```
I think https://www.ncbi.nlm.nih.gov/pmc/articles/PMC3596837/ is a better reference.

pc494 commented on pyxem/utils/indexation_utils.py:
```diff
@@ -90,9 +172,25 @@ def correlate_library(image, library, n_largest, mask):
+
     top_matches = np.empty((len(library), n_largest, 3), dtype='object')
 
+    variables_dict = {}
```
this is unused I think?

pc494 commented on pyxem/tests/test_utils/test_indexation_utils.py:
```diff
@@ -20,7 +20,34 @@
+
+
+def test_normalized_correlation():
+    nb_pixels = 3
```
Would be happy for these to just read:
`assert np.allclose(normalized_correlation(3,1,1/3,[1,0,0],[1,0,0]),1)`

Not sure explicitly calculating helps very much and just adds dead line (also "mean" is preferred to "average")

EirikOpheim commented:
I believe I have made all the requested changes now, Phillip! Thanks for very thorough code reviews. I believe I will be able to write closer to the pyxem standard in my next PR :)

pc494 commented:
@EirikOpheim I will make the final changes to this and then merge it. Good work.

pc494 commented:
Once https://github.com/EirikOpheim/pyxem/pull/3 is in I'll merge this.

dnjohnstone commented:
Well done @EirikOpheim - this is a really solid early contribution! Thanks @pc494 for all the great reviewing.

EirikOpheim commented:
> Looks good, onwards we head!

Yea!

> Well done @EirikOpheim - this is a really solid early contribution!

Thanks! :)
